replica/table.cc

```diff
@@ -741,29 +741,34 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
     if (_async_gate.is_closed() || is_auto_compaction_disabled_by_user()) {
         co_return;
     }
-    auto sstable_count_below_threshold = [this] {
-        const auto sstable_runs_with_memtable_origin = boost::copy_range<std::unordered_set<utils::UUID>>(
-            *_main_sstables->all()
+    auto num_runs_with_memtable_origin = [this] {
+        auto& cm = _compaction_manager;
+        auto& cs = get_compaction_strategy();
+        auto desc = cs.get_sstables_for_compaction(as_table_state(), cm.get_strategy_control(), cm.get_candidates(*this));
```

Comment:
What if this chooses sstables that aren't memtable-origin for the "best" compaction? Then we get a false negative.

Probably it's not terrible. If the best compaction happens not to include memtable-origin sstables (or not all of them) then the memtable-origin sstables probably aren't a problem.

But we may still have a problem with filtering only on memtable-origin sstables: the first compaction converts memtable-origin sstables to non-memtable-origin sstables, and if that compaction is fast enough but the following compactions aren't, we may still be creating sstables faster than we're compacting them. It's probably unlikely to happen though.

We may want to drop the whole memtable-origin stuff and instead just ask the compaction strategy if it's overloaded and if so stop flushing new memtables. Exactly what "overloaded" means will be left to the compaction strategy to decide.

We could do this later though.

Reply:
> What if this chooses sstables that aren't memtable-origin for the "best" compaction? Then we get a false negative.
> 
> Probably it's not terrible. If the best compaction happens not to include memtable-origin sstables (or not all of them) then the memtable-origin sstables probably aren't a problem.

Agreed.

> But we may still have a problem with filtering only on memtable-origin sstables: the first compaction converts memtable-origin sstables to non-memtable-origin sstables, and if that compaction is fast enough but the following compactions aren't, we may still be creating sstables faster than we're compacting them. It's probably unlikely to happen though.
> 
> We may want to drop the whole memtable-origin stuff and instead just ask the compaction strategy if it's overloaded and if so stop flushing new memtables. Exactly what "overloaded" means will be left to the compaction strategy to decide.
> 
> We could do this later though.

The "compaction overloaded" metric is similar to the compaction backlog.
The difference is that the backlog still doesn't take into account the number of sstables, only their size (don't be mislead by #4076 being closed), so we don't factor in potential read-amplification.

Reply:
It doesn't have to be similar to compaction backlog. It can be a pure sstable count thing.